ppafm/core.py

```diff
@@ -150,14 +150,30 @@ def setTip(lRadial=None, kRadial=None, rPP0=None, kSpring=None, parameters=None)
     lib.setTip(lRadial, kRadial, rPP0, kSpring)
 
 
-# void setTipSpline( int n, double * xs, double * ydys ){
-lib.setTipSpline.argtypes = [c_int, array1d, array2d]
-lib.setTipSpline.restype = None
-
-
-def setTipSpline(xs, ydys):
-    n = len(xs)
-    lib.setTipSpline(n, xs, ydys)
+# struct SplineParams {
+#     int      rff_n;
+#     double * rff_xs;
+#     double * rff_ydys;
+# }
+class SplineParameters(Structure):
+    _fields_ = [
+        ("rff_n", c_int),
+        ("rff_xs", POINTER(c_double)),
+        ("rff_ydys", POINTER(c_double)),
+    ]
+
+    def __init__(self, xs, ydys):
+        super().__init__()
+        self.np_rff_xs = np.ascontiguousarray(xs, dtype=np.double)
+        self.np_rff_ydys = np.ascontiguousarray(ydys, dtype=np.double)
+        self.rff_n = len(self.np_rff_xs)
+        self.rff_xs = np.ctypeslib.as_ctypes(self.np_rff_xs)
+        self.rff_ydys = np.ctypeslib.as_ctypes(self.np_rff_ydys.reshape(-1))
+
+    @classmethod
```

Comment:
I did not know this pattern `@classmethod` before. Basically it is like overloaded constructor (?)

Reply:
Kind of, yes. It makes the method so that it can be called without creating an instance of the class, and the first argument, instead of being `self` pointing to the instance, is `cls` which points to the class constructor.